add token exchange proxy via `POST /api/storage/s3-token`

overheadhunter opened the pull request:
This PR adds a new API endpoint to the API for token exchange, making it available to public clients (if they possess a token with an appropriate audience).

## Changes done in Keycloak:

* reconfigured `cryptomatorvaults`, adding a non-empty secret and removing some default attributes

## Changes done in Hub

1. Added `POST /api/storage/s3-token` endpoint, which will make the actual token exchange request using the credentials of `cryptomatorvaults`
2. Added integration test which first requests an access token for `cryptomator`, then exchanges it for an access token for `cryptomatorvaults`

## Misc

* Added test dependencies
* Sadly, I had to enable Keycloak dev-services for unit tests (formerly mocked), slowing down the build. Maybe there is a better way to run integration tests against a real KC instance?
* Fixes incorrect name of dev-realm.json (apparently, newer KC versions complain, if the file name doesn't match the realm name)
* Removed ant-run plugin that renamed (no longer existing) dev-realm.json, preventing tests from executing

## Protocol Changes

Step 7 and 8 in the following diagram are now proxied, so the `client_id` and `client_secret` of the confidential client (i.e. Hub backend) can be used to perform the token exchange:

```mermaid
sequenceDiagram
    participant User
    participant Katta Client
    participant Katta Server
    participant Keycloak
    participant S3
    Katta Client ->> Keycloak: (1) /authorize
    Keycloak -->> Katta Client: (2) redirect
    Katta Client --) User: (2) prompt in browser
    User --) Keycloak: (3) User/Password
    Keycloak --) Katta Client: (4) Authorization Code

    Katta Client ->> Keycloak: (5): /token
    Note over Keycloak: client_id: cryptomator
    Keycloak -->> Katta Client: (6) access_token: OIDC Access Token (JWT)

    Note over Katta Client: { aud: "cryptomator", ... }

    Katta Client ->> Katta Server: (7a) POST /api/storage/s3-token<br/>Authorization: Bearer fo0b4r==
    Katta Server ->> Keycloak: (7b): POST realms/{realm}/protocol/openid-connect/token<br/>Authorization: Basic cryptomatorvaults:client_secret
    Note over Keycloak: grant_type: token-exchange<br/> subject_token: <access_token><br/> scope: <vaultId>
    Keycloak -->> Katta Server: (8a) exchanged_access_token: OIDC Access Token (JWT)
    Katta Server ->> Katta Client: (8b) exchanged_access_token: OIDC Access Token (JWT)

    Note over Katta Client: { aud: "cryptomatorvaults", <br/>azp: "cryptomatorvaults, ... }

    Katta Client ->> S3: 

```

overheadhunter committed:
add `POST /api/storage/s3-token`

to exchange token issued for `cryptomator` public client for `cryptomatorvaults` client

overheadhunter committed:
token exchange officially supported since KC 26.2

see https://www.keycloak.org/2025/04/keycloak-2620-released

overheadhunter committed:
token exchange officially supported since KC 26.2

see https://www.keycloak.org/2025/04/keycloak-2620-released

overheadhunter committed:
token exchange officially supported since KC 26.2

see https://www.keycloak.org/2025/04/keycloak-2620-released

overheadhunter committed:
disable needless S3 dev services

because localstack container requires rootful container execution,
which is disabled by default in podman, unless activated via `podman machine set --rootful=true`

also speeding up builds a bit

overheadhunter commented on backend/src/main/java/org/cryptomator/hub/api/cipherduck/StorageResource.java:
```diff
@@ -88,4 +93,20 @@ public Response createBucket(@PathParam("vaultId") UUID vaultId, final CreateS3S
+				jwt.getRawToken(),
+				"urn:ietf:params:oauth:token-type:access_token",
+				"urn:ietf:params:oauth:token-type:access_token",
+				"email", // TODO: parametrize scope?
```
probably should contain the vault id

chenkins replied:
yes.

overheadhunter replied:
changed in 2d881b6

chenkins commented on backend/src/test/java/org/cryptomator/hub/cipherduck/KeycloakContainerResource.java:
```diff
@@ -16,7 +14,7 @@ public Map<String, String> start() {
 				//				.withDebugFixedPort(5005, false)
 				//				.withCustomCommand("--log-level=DEBUG")
-				.withRealmImportFile("/dev.json");
+				.withRealmImportFile("/cryptomator-realm.json");
```
Let's minimize diff to upstream, this is already painful. I suggest doing the renaming upstream.

There is already a working solution: https://github.com/shift7-ch/katta-server/blob/feature/cipherduck-uvf/backend/pom.xml#L289-L291. IDE integration is not great, yes (manually run `mvn clean verify -DskipTests` to generate the renamed file).

chenkins commented on backend/src/main/resources/application.properties:
```diff
@@ -28,10 +28,11 @@ quarkus.oidc.application-type=service
 
 # Keycloak dev service
-%dev.quarkus.keycloak.devservices.realm-path=dev-realm.json
+%dev.quarkus.keycloak.devservices.realm-path=cryptomator-realm.json
```
Dito, see below.

chenkins commented on keycloak/src/test/java/cloud/katta/KattaTokenExchangeIT.java:
I don't think we should discard this test even if we hide Keycloak Token-Exchange behind an OAuth gateway. I would want to know before a version upgrade that Keycloak behaves the same.

overheadhunter replied:
reverted in e69ac9b

chenkins commented on backend/src/test/java/org/cryptomator/hub/cipherduck/TokenExchangeIT.java:
```diff
@@ -0,0 +1,67 @@
+	@DisplayName("authenticate as public client 'cryptomator' and exchange token for 'cryptomatorvaults'")
+	public void testTokenExchange() throws GeneralSecurityException, IOException, InterruptedException {
+		// 1. Authenticate as public client using Authorization Code Flow with PKCE
+		var authResponse = TinyOAuth2.client("cryptomator") //
```
This is better than enabling PasswordGrant, nice.

chenkins reviewed (changes requested):
to be discussed f2f

overheadhunter committed:
Revert aa5925e, 2309ba3, df400cf

overheadhunter committed:
`dev-realm.json` → `cryptomator-realm.json`

overheadhunter committed:
exchange token as `cryptomatorvaults`

making sure `azp` header is `cryptomatorvaults`

```
{
  "exp": 1752145041,
  "iat": 1752144741,
  "auth_time": 1752144741,
  "jti": "ntrtte:5f48bed6-906e-4078-a7d1-63ff6ebef7da",
  "iss": "http://localhost:37965/realms/cryptomator",
  "sub": "758babab-ec50-4d0d-ae2d-828625e54fba",
  "typ": "Bearer",
  "azp": "cryptomatorvaults",
  "sid": "8f12aa49-f138-4664-9d2e-1ba3a520b30e",
  "scope": "address",
  "address": {}
}
```

overheadhunter committed:
exchange token as `cryptomatorvaults`

making sure `azp` header is `cryptomatorvaults`

```
{
  "exp": 1752145041,
  "iat": 1752144741,
  "auth_time": 1752144741,
  "jti": "ntrtte:5f48bed6-906e-4078-a7d1-63ff6ebef7da",
  "iss": "http://localhost:37965/realms/cryptomator",
  "sub": "758babab-ec50-4d0d-ae2d-828625e54fba",
  "typ": "Bearer",
  "azp": "cryptomatorvaults",
  "sid": "8f12aa49-f138-4664-9d2e-1ba3a520b30e",
  "scope": "address",
  "address": {}
}
```

overheadhunter committed:
parameterized: `/api/storage/s3-token?vault=foo`

overheadhunter committed:
adjust credentials

overheadhunter committed:
re-add `"aud": "cryptomatorvaults"`:

```
{
  "exp": 1752147102,
  "iat": 1752146802,
  "auth_time": 1752146802,
  "jti": "ntrtte:2377d26d-12e7-4f97-9afb-7bcfc23313f3",
  "iss": "http://localhost:34455/realms/cryptomator",
  "aud": "cryptomatorvaults",
  "sub": "e3545ec6-4421-4ca6-858a-3b77000043cc",
  "typ": "Bearer",
  "azp": "cryptomatorvaults",
  "sid": "812a1d9b-342a-4cfb-b8b3-d3c5824bc08f",
  "scope": "address",
  "address": {}
}
```

overheadhunter committed:
adjust credentials

overheadhunter committed:
reordered lines to minimize diff
